steps/expval.py

```diff
@@ -0,0 +1,159 @@
+# Copyright 2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+The PennyLane-Orquestra step for computing the expectation value of Hermitian
+operators given a quantum circuit and a quantum device.
+
+Functions defined in this file may be included in an Orquestra workflow file as
+a workflow step. Such workflow steps are executed on a remote Orquestra node.
+"""
+import json
+
+import numpy as np
+from openfermion import IsingOperator, QubitOperator
+from qiskit import QuantumCircuit
+
+from zquantum.core.circuit import Circuit
+from zquantum.core.measurement import expectation_values_to_real
+from zquantum.core.utils import create_object, save_list
+
+
+def run_circuit_and_get_expval(
+    backend_specs: str,
+    circuit: str,
+    operators: str,
+):
+    """Takes a circuit to obtain the expectation value of an operator on a
+    given backend.
+
+    All Orquestra backend interface calls used in this function are standard
+    methods of the ``QuantumBackend`` and ``QuantumSimulator`` interfaces as
+    defined in the ``z-quantum-core`` repository.
+
+    There are two computation modes: sampling and exact. Expectation values are
+    computed by post-processing samples when an Orquestra ``QuantumBackend`` is
+    used or the number of samples was specified for a ``QuantumSimulator``
+    backend. When the number of samples isn't specified, ``QuantumSimulator``
+    backends run in exact mode.
+
+    Args:
+        backend_specs (str): the Orquestra backend specification in a json
+            representation
+        circuit (str): the circuit represented as an OpenQASM 2.0 program
+        operators (str): the operator in an ``openfermion.QubitOperator``
+            or ``openfermion.IsingOperator`` representation
+    """
+    backend_specs = json.loads(backend_specs)
+    operators = json.loads(operators)
+
+    backend = create_object(backend_specs)
+
+    # 1. Parse circuit
+    qc = QuantumCircuit.from_qasm_str(circuit)
+
+    # 2. Create operators
+    ops = []
+    if backend.n_samples is not None:
+        # Operators for Backend/Simulator in sampling mode
+        for op in operators:
+            ops.append(IsingOperator(op))
+    else:
+        # Operators for Simulator exact mode
+        for op in operators:
+            ops.append(QubitOperator(op))
```

Comment:
This is how Orquestra computes expectation values internally:
```python
        # We require operator to be IsingOperator because measurements are always performed in the Z basis,
        # so we need the operator to be Ising (containing only Z terms).
        # A general Qubit Operator could have X or Y terms which don’t get directly measured.
        if isinstance(ising_operator, IsingOperator) == False:
            raise Exception("Input operator is not openfermion.IsingOperator")
```
For the [`get_expectation_values` function that is called on samples](https://github.com/zapatacomputing/z-quantum-core/blob/master/src/python/zquantum/core/measurement.py).

Reply:
Huh. I find that a super strange restriction; why not just check if the IsingOperator _or_ QubitOperator consists of only `Z`.

Reply:
Hmm, I vaguely remember as if it was required due to how openfermion is being used at times. There's a convenience function used at times to convert between types of openfermion operators:
https://github.com/zapatacomputing/qe-openfermion/blob/6d52df917727dc71883b5672ce8015fd917a6d98/src/python/qeopenfermion/_utils.py#L350